File: Algorithmique appliquée/chapitre2/cpp/exercice6.cpp
/* programme facteursPremiers (exercice 7)
*  but : saisir des valeurs, afficher les facteurs premiers, puis au final
*        afficher une seule fois chaque facteur premier qui est apparu
*        Ceci est la version optimisée
*  date : 26/10/2016
*/
#include <iostream>
using namespace std;


int main(int argc, char *argv[]){
    unsigned int facteur[100];
    unsigned int valeur, diviseur;
    unsigned int i,j;
    unsigned int maxM;


    cout << "entrer une valeur (0 pour sortir) =";
    cin >> valeur;
    while (valeur != 0){
        diviseur =2;
        while(diviseur <= 0){
            // il s'agit d'un nombre premier
            if(valeur%diviseur ==0){
                //on l'affiche
                cout << diviseur << " ";
                i =0;
                while (i < maxM && facteur[i]!=diviseur){
                    i++;
                }
                if(i==maxM){
                    facteur[maxM++] = diviseur;
                }
                valeur /= diviseur;
            }else{
                diviseur++;
            }
        }
        cout << endl << "Tous les facteurs premiers = ";
        cin >> valeur;
    }

    cout << endl << "tous les facteurs premiers =" << " ";
    for(j=0;j < maxM;j++){
        cout << facteur[j] << " ";
    }

    return 0;
}

File: Algorithmique appliquée/chapitre2/python/exercice6.py
# -*-coding:utf-8 -*
# Programme exercice 6
# Saisir des valeurs, afficher les facteurs premiers, puis au final
# afficher une seule fois chaque facteur premier qui est apparu
# Ceci est la version optimisée
# date : 26/10/2016


import os


# déclarations
facteur = [0]*100
max = 0 # nbre de cases remplies dans facteur

# boucle sur la saisie des valeurs
valeur = int(input("entrer une valeur (0 pour finir) = "))
while valeur != 0 :
    # recherche des facteurs premiers
    diviseur = 2
    while diviseur <= valeur :
        # si on trouve un facteur premier
        if valeur % diviseur == 0 :
            # on l'affiche
            print(diviseur, " ")
            # on regarde s'il a déjà été mémorisé
            k = 0
            while k < max and facteur[k] != diviseur :
                k += 1
            if k == max :
                facteur[max] = diviseur
                max += 1
            # on divise la valeur pour éviter les facteurs non premiers
            valeur /= diviseur
        else :
            diviseur += 1

    # saisie d'une nouvelle valeur
    valeur = int(input("entrer une valeur (0 pour finir) = "))

# affichage une seule fois de chaque facteur premier
print("Tous les facteurs premiers = ")
for k in range(0, max) :
    print(facteur[k])


os.system("pause")
